### PR TITLE
[MPS][TYPE_PROMOTION] Fix Clamp

### LINKED ISSUE #130201: tensor.clamp produces wrong values on MPS
### 🐛 Describe the bug

I encountered some weird NaNs when switching from cpu to mps and after a bit of digging, noticed that clamp behaves differently. The issue is reproducible from the python console:
```python
import torch
torch.tensor(0).clamp(min=1e-1)  # tensor(0.1000)
torch.tensor(0).to("mps").clamp(min=1e-1).item(). # 0.0
```

image

### Versions

Collecting environment information...
PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.28.3
Libc version: N/A

Python version: 3.10.4 (main, Sep 11 2022, 20:39:17) [Clang 13.1.6 (clang-1316.0.21.2.5)] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] numpy-quaternion==2023.0.2
[pip3] pytorch-lightning==2.2.0.post0
[pip3] torch==2.3.1
[pip3] torch-tb-profiler==0.4.3
[pip3] torchmetrics==1.3.1
[pip3] torchvision==0.18.1
[conda] No relevant packages

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130226

Summary:
1. Fixed #130201 by adding type promotion.
2. Added proper tests.
3. Found torch's type promotion is different from numpy as follows:

```python
import torch
import numpy as np
np.clip(np.array([1], dtype=np.float32), np.array([1], dtype=np.int32), None).dtype  # dtype('float64')
torch.clamp(torch.tensor([1], dtype=torch.float32), torch.tensor([1], dtype=torch.int32)).dtype  # torch.float32
```

~Not sure the proper way to handle it, it causes numpy ref tests to fail.~
Reason here, so think I'm gonna xfail it:
https://github.com/pytorch/pytorch/blob/3c1cf03fde145bdbe1f5ffb81765d076c10b4c04/test/test_ops.py#L260-L264

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/TensorCompare.mm [hunk 1]
```diff
@@ -238,21 +237,23 @@ static void clamp_scalar_out_mps(const Tensor& input_t,
...
     string key = op_name + (has_min ? ("_min:" + std::to_string(min_scalar)) : "") +
         (has_max ? ("_max:" + std::to_string(max_scalar)) : "") + "_scalar:" + getTensorsStringKey({input_t});
     auto cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>(key, [&](auto mpsGraph, auto newCachedGraph) {
       if (has_min)
-        newCachedGraph->minTensor = [mpsGraph
-            constantWithScalar:min_scalar
-                         shape:(mps::getMPSShape(input_t))dataType:(mps::getMPSScalarType(input_t.scalar_type()))];
+        newCachedGraph->minTensor =
+            [mpsGraph constantWithScalar:min_scalar
+                                   shape:(mps::getMPSShape(input_t))dataType:(mps::getMPSScalarType(result_type))];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/TensorCompare.mm:250]:
```suggestion
                                   shape:mps::getMPSShape(input_t)
                                   dataType:mps::getMPSScalarType(result_type)];
```

File: aten/src/ATen/native/mps/operations/TensorCompare.mm [hunk 2]
```diff
@@ -238,21 +237,23 @@ static void clamp_scalar_out_mps(const Tensor& input_t,
...
       if (has_max)
-        newCachedGraph->maxTensor = [mpsGraph
-            constantWithScalar:max_scalar
-                         shape:(mps::getMPSShape(input_t))dataType:(mps::getMPSScalarType(input_t.scalar_type()))];
+        newCachedGraph->maxTensor =
+            [mpsGraph constantWithScalar:max_scalar
+                                   shape:(mps::getMPSShape(input_t))dataType:(mps::getMPSScalarType(result_type))];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/TensorCompare.mm:254]:
```suggestion
                                   shape:mps::getMPSShape(input_t)
                                   dataType:mps::getMPSScalarType(result_type)];
```

**Review by malfet — APPROVED** [review]

**Comment by qqaatw** [comment]
@malfet can we cherry-pick this PR to v2.4.1. People are reporting issues: #133100

**Pull request closed without merging**